### ISSUE TITLE
Order of operations issue with Dependency registrations

### ISSUE THREAD

**Issue by scionwest** [PrismLibrary/Prism#877]
- Platform: Xamarin Forms
- Prism version: Prism.Forms

### Repro steps

Using the Prism.Autofac.Forms library, I'd like to see adjustments to how the base class handles registration types. For example, I have a Xamarin.Forms project with an Android and iOS app with a PCL library. The App.cs file in my PCL inherits from `PrismApplication` and is provided with a custom interface called `IDependencyRegistration`.

```C#
    public interface IDependencyRegistration
    {
        void RegisterDependencies();
    }

    public class App : PrismApplication
    {
        private IDependencyRegistration dependencyRegistration;

        public App(IDependencyRegistration dependencyRegistration)
        {
            this.dependencyRegistration = dependencyRegistration;
        }

        protected override void OnInitialized()
        {
            MobileCenter.Start(typeof(Analytics), typeof(Crashes));
            this.NavigationService.NavigateAsync("MainPage");
        }

        protected override void RegisterTypes()
        {
            this.dependencyRegistration?.RegisterDependencies();
        }
    }
```

This allows each platform to register dependency implementations that are platform specific.

```c#
    internal class AndroidDependencyService : IDependencyRegistration
    {
        public void RegisterDependencies()
        {
            DependencyService.Register<AndroidSqliteProvider>();
        }
    }

    public class MainActivity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
    {
        protected override void OnCreate(Bundle bundle)
        {
            // ......
            var app = new App(new AndroidDependencyService());
            LoadApplication(app);
        }
    }
```

With the current implementation of the `PrismApplication`, my constructor is not allowed to run, before the `RegisterTypes` method is called. This prevents me from running Type registration stuff in that method. I know I could easily just call `dependencyRegistration.RegisterDependencies()` from within my constructor, but then my code has platform registration and shared registration happening in two different places.

Looking at [PrismApplicationBase](https://github.com/PrismLibrary/Prism/blob/master/Source/Xamarin/Prism.Forms/PrismApplicationBase.cs#L46) it looks like you initialize the app in your constructor, which eventually leads to invoking `RegisterTypes`. I'd propose moving the initialization out of the constructor (giving sub-classes a chance to configure themselves) and into the `OnStart` method of the base Xamarin `Application` class.

```C#
        protected override void OnStart() 
        { 
                InitializeInternal();
        } 
```

This allows sub-classed Apps to take dependencies from each platform, configure them if needed, and then use them during the actual Prism initialization process.

**Comment by scionwest** [comment]
It looks like you actually already addessed this with #630. I've been piecing it together from the unit tests (can't find docs on autofac's flavor) and I keep hitting a runtime `NullReferenceException`. Is there something I'm doing wrong with my setup? The exception is happening outside of my code (looks like in `platform.cs` in Xamarin).

### App.cs

```c#
    public class App : PrismApplication
    {
        public App(IPlatformInitializer initializer) : base(initializer) { }

        protected override void OnInitialized()
        {
            this.NavigationService.NavigateAsync("MainPage");
        }

        protected override void RegisterTypes()
        {
            var builder = new ContainerBuilder();
            builder.RegisterType<MainPage>();
            builder.Update(base.Container);
        }
    }
```

### MainActivity

```c#
    public class AndroidConfiguration : IPlatformInitializer
    {
        public void RegisterTypes(IContainer container)
        {
            var builder = new ContainerBuilder();
            builder.RegisterType<AndroidSqliteProvider>().AsImplementedInterfaces();
            builder.Update(container);
        }
    }

    [Activity(Label = "MobileApp", Icon = "@drawable/icon", Theme = "@style/MainTheme", MainLauncher = true, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation)]
    public class MainActivity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
    {
        protected override void OnCreate(Bundle bundle)
        {
            TabLayoutResource = Resource.Layout.Tabbar;
            ToolbarResource = Resource.Layout.Toolbar;

            base.OnCreate(bundle);

            global::Xamarin.Forms.Forms.Init(this, bundle);
            var app = new App(new AndroidConfiguration());
            LoadApplication(app);
        }
    }
```

The exception is thrown once the `OnCreate(Bundle bundle)` method is completed.

### MainPage.xaml

```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:MobileApp"
             xmlns:prism="clr-namespace:Prism.Mvvm;assembly=Prism.Forms"
             prism:ViewModelLocator.AutowireViewModel="True"
             x:Class="MobileApp.Views.MainPage">
  
  <Frame OutlineColor="Gray" 
         HorizontalOptions="Center" 
         VerticalOptions="Center">
    <Label Text="{Binding Title}"
           VerticalOptions="Center"
           HorizontalOptions="Center" />
  </Frame>
  
</ContentPage>
```

### MainPageViewModel

```c#
    public class MainPageViewModel : BindableBase
    {
        private string title = "Hello from VM";
        public string Title
        {
            get
            {
                return title;
            }

            set
            {
                SetProperty(ref title, value);
            }
        }
    }
}
```

### Callstack

>	0x95 in Xamarin.Forms.Platform.Android.AppCompat.Platform.LayoutRootPage at C:\BuildAgent\work\aad494dc9bc9783\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:279,5	C#
 	0x12 in Xamarin.Forms.Platform.Android.AppCompat.Platform.Xamarin.Forms.Platform.Android.IPlatformLayout.OnLayout at C:\BuildAgent\work\aad494dc9bc9783\Xamarin.Forms.Platform.Android\AppCompat\Platform.cs:196,5	C#
 	0x20 in Xamarin.Forms.Platform.Android.PlatformRenderer.OnLayout at C:\BuildAgent\work\aad494dc9bc9783\Xamarin.Forms.Platform.Android\PlatformRenderer.cs:73,4	C#
 	0x12 in Android.Views.ViewGroup.n_OnLayout_ZIIII at /Users/builder/data/lanes/3511/ce955cc0/source/monodroid/src/Mono.Android/platforms/android-23/src/generated/Android.Views.ViewGroup.cs:3427,4	C#

Any ideas on what I'm doing wrong here? The `MainPage` and `MainPageViewModel` bindings were working fine until I brought in and configured the `Prism.Autofac.Forms` package. It's really got me confused.

**Comment by brianlagunas** [comment]
Sorry, I have zero experience with autofac.  Could it e related to #875? If not, maybe someone form the community can help.

**Comment by scionwest** [comment]
Prism comes with support for it out of the box though, Prism.Autofac.Forms is your package. Couldn't you at least tell me if I was using the Prism part of it correctly? I don't expect you to support the Autofac API itself, but Prisms registration process is what I'm having issues with.

**Comment by brianlagunas** [comment]
Prism.Autofac.Forms was added by the community via a PR, and therefore must be supported by the community.  If the expectation is that if I ship a DI container package, then I am responsible for supporting the functionality of the DI container, then I will delete all DI packages immediately.  I am not about to drag myself into that support lifecycle :)

As far as the IPlatformInitializer, yes you are using it right.  You register your platform specific classes to the interfaces that are defined in the PCL.  Though, I am not familiar with how you register types with Autofac via it's container so I can't help there.  Looking at your code, it doesn't look like that is even related because you are not trying to resolve the AndroidSqliteProvider yet.  So the issue lies somewhere else.  Possibly in just creating the view itself.

If you look through the issue I linked to, it leads to a code snippet that may or may not be related to your issue.  Please see if it fixes your issue:

https://github.com/PrismLibrary/Prism/issues/742#issuecomment-263822188

**Comment by scionwest** [comment]
Yeah, i'm pretty sure it's something related to instantiating the view itself. Not sure what's happening. Debugging these mobile apps is not nearly as easy as debugging a WPF app!

**Comment by dansiegel** [comment]
@scionwest if you want to put a link up for your app I'd be happy to take a look. I know SQLite was one of the reasons that IPlatformInitializer was added since you have to register the implementation with the container on the specific platform and it is called before ```OnInitialized``` so that the first time you navigate, your dependencies are available.

**Comment by bartlannoeye** [comment]
The code linked in #742 is not in the latest pre-release package yet. I hope to find some time one of the upcoming days to push a new prelease package. If you could post the code to help us reproduce the issue, I'll test that against the lat codebase as well.

**Comment by scionwest** [comment]
@bartlannoeye I [pushed a repro](https://github.com/scionwest/Prism.Forms.AutofacBug) up this evening for you - bare bones solution for you to test against.

**Comment by bartlannoeye** [comment]
Thanks, I'll look into it asap. A bit under the weather right now.

**Comment by brianlagunas** [comment]
What's the status of this issue?

**Comment by brianlagunas** [comment]
So I took a look into this and found that it is in fact related to the issues linked previously.  Autofac does not automatically resolve concrete classes that have not been registered.  This means that the MainPageViewModel could not be resolved and was throwing an exception.  A work around is to add this code snippet as the last code block in RegisterTypes:
```
        protected override void RegisterTypes()
        {
            base.Container.RegisterTypeForNavigation<MainPage, MainPageViewModel>("MainPage");

            var containerUpdater = new ContainerBuilder();
            containerUpdater.RegisterSource(new AnyConcreteTypeNotAlreadyRegisteredSource());
            containerUpdater.Update(Container);
        }
```
This means that this should be fixed with the next preview.

Now having said that, I did notice that the RegisterTypeForNavigation is only registering the View and not the ViewModel.  Not sure why that was done, but we could probably automatically register the VM as well in that method.
